build/xtask/src/dist.rs

```diff
@@ -1712,198 +1711,156 @@ fn allocate_one(
-        attributes: abi::RegionAttributes::empty(), // no rights
-    });
+) -> Result<build_kconfig::KernelConfig> {
+    let mut tasks = vec![];
+    let mut irqs = BTreeMap::new();
 
-    // Regions 1.. are the fixed peripheral regions, shared by tasks that
-    // reference them. We'll build a lookup table so we can find them
-    // efficiently by name later.
-    let mut peripheral_index = IndexMap::new();
-    let sname = &"secure".to_string();
+    let p2_required = toml.mpu_power_of_two_required();
 
-    // Build a set of all peripheral names used by tasks, which we'll use
-    // to filter out unused peripherals.
-    let used_peripherals = toml
-        .tasks
-        .iter()
-        .flat_map(|(_name, task)| task.uses.iter())
-        .collect::<HashSet<_>>();
-
-    let power_of_two_required = toml.mpu_power_of_two_required();
-    for (name, p) in toml.peripherals.iter() {
-        if power_of_two_required && !p.size.is_power_of_two() {
-            panic!("Memory region for peripheral '{}' is required to be a power of two, but has size {}", name, p.size);
-        }
-
-        // skip periperhals that aren't in at least one task's `uses`
-        if !used_peripherals.contains(name) {
-            continue;
+    let mut flat_shared = BTreeMap::new();
+    for (name, p) in &toml.peripherals {
+        if p2_required && !p.size.is_power_of_two() {
+            bail!(
+                "memory region for peripheral '{}' is required to be \
+                   a power of two, but has size {}",
+                name,
+                p.size
+            );
         }
 
-        peripheral_index.insert(name, regions.len());
-
-        // Peripherals are always mapped as Device + Read + Write.
-        let attributes = abi::RegionAttributes::DEVICE
-            | abi::RegionAttributes::READ
-            | abi::RegionAttributes::WRITE;
-
-        regions.push(abi::RegionDesc {
-            base: p.address,
-            size: p.size,
-            attributes,
-        });
+        flat_shared.insert(
+            name.to_string(),
+            build_kconfig::RegionConfig {
+                base: p.address,
+                size: p.size,
+                attributes: build_kconfig::RegionAttributes {
+                    read: true,
+                    write: true,
+                    execute: false,
+                    special_role: Some(build_kconfig::SpecialRole::Device),
+                },
+            },
+        );
     }
-
-    for (name, p) in toml.extratext.iter() {
-        if power_of_two_required && !p.size.is_power_of_two() {
-            panic!("Memory region for extratext '{}' is required to be a power of two, but has size {}", name, p.size);
+    for (name, p) in &toml.extratext {
+        if p2_required && !p.size.is_power_of_two() {
+            bail!(
+                "memory region for extratext '{}' is required to be \
+                   a power of two, but has size {}",
```

Comment:
Nit: weird indentation

Reply:
Huh, interesting, I did not write it that way -- rustfmt's wrapped string handling seems hard to predict.